Scripts/Runtime/Entities/Lifecycle/EntitySpawnSystem.cs

```diff
@@ -171,44 +151,216 @@ public void UnregisterEntityPrototypeForDefinition<TEntitySpawnDefinition>(bool
+        // SPAWN API - IN JOB
         //*************************************************************************************************************
+        
+        /// <summary>
+        /// Acquires an <see cref="EntitySpawner"/> for use in a job and returns the dependency to wait on.
+        /// Must call <see cref="ReleaseAsync"/> with the dependency for the jobs that use this instance.
+        /// </summary>
+        /// <param name="value">The <see cref="EntitySpawner"/> instance</param>
+        /// <returns>The <see cref="JobHandle"/> to wait on</returns>
+        public JobHandle AcquireAsync(out EntitySpawner value)
+        {
+            JobHandle prototypesDependency = m_EntityPrototypes.AcquireAsync(AccessType.SharedRead, out var prototypes);
+            value = AcquireSpawner(true, prototypes);
+            return prototypesDependency;
+        }
 
-        protected override void OnUpdate()
+        /// <summary>
+        /// Lets this system know of jobs that will use an acquired <see cref="EntitySpawner"/> via
+        /// <see cref="AcquireAsync"/> so that the <see cref="EntityCommandBufferSystem"/> can ensure the jobs are
+        /// complete before playing back the underlying <see cref="EntityCommandBuffer"/>
+        /// </summary>
+        /// <param name="releaseAccessDependency"></param>
+        public void ReleaseAsync(JobHandle releaseAccessDependency)
         {
-            Dependency = ScheduleActiveEntitySpawners(Dependency);
+            m_EntityPrototypes.ReleaseAsync(releaseAccessDependency);
+            m_CommandBufferSystem.AddJobHandleForProducer(releaseAccessDependency);
+        }
 
-            //Ensure we're turned back off
-            m_ActiveEntitySpawners.Clear();
-            Enabled = false;
+        
+        private EntitySpawner AcquireDeferred()
+        {
+            return AcquireSpawner(true, m_EntityPrototypes.Acquire(AccessType.SharedRead));
         }
 
-        private JobHandle ScheduleActiveEntitySpawners(JobHandle dependsOn)
+        private void ReleaseDeferred(in EntitySpawner entitySpawner)
         {
-            NativeArray<JobHandle> dependencies = new NativeArray<JobHandle>(m_ActiveEntitySpawners.Count, Allocator.Temp, NativeArrayOptions.UninitializedMemory);
+            m_EntityPrototypes.Release();
+            ReleaseSpawner(entitySpawner);
+        }
 
-            //All the active spawners that need to go this frame are given a chance to go ahead and run their spawn jobs
-            int index = 0;
+        private EntitySpawner AcquireImmediate()
+        {
+            return AcquireSpawner(false, m_EntityPrototypes.Acquire(AccessType.SharedRead));
+        }
 
-            foreach (IEntitySpawner entitySpawner in m_ActiveEntitySpawners)
+        private void ReleaseImmediate(in EntitySpawner entitySpawner)
+        {
+            m_EntityPrototypes.Release();
+            entitySpawner.Playback(EntityManager);
+            entitySpawner.DisposeECB();
```

Comment:
I'm not sure we should be playing back immediately on an immediate release. Shouldn't we still be waiting for the ECB playback moment that this `EntitySpawnSystem` is configured to execute?

If you need a job handle for `m_CommandBufferSystem.AddJobHandleForProducer` you can just pass default since your "work" with the ECB is done.

Reply:
Oh I see, this isn't actually an ACV compatible type. Is there a way to require these in jobs without setting up a custom `Schedule` function?

Would really like to maintain compatibility with the "Require" pattern.

Reply:
These functions are private. 

The public facing ones are 

`public JobHandle AcquireAsync(out EntitySpawner value)`
`public void ReleaseAsync(JobHandle releaseAccessDependency)`

I'm ok to implement one of the `IAccessController` interfaces but they add in a bunch of extra stuff like `Dispose` and the synchronous `Acquire` functions that don't quite make sense. 

I could explicitly implement those functions and just toss a `NotSupported` exception but that does feel a bit dirty.

Reply:
`IAccessController` doesn't have `Dispose`

For the sync interface there's no real benefit over SpawnImmediate but no a problem to support either. Maybe you want to schedule an entity to exist at the system's sync point and not until then.

Reply:
Ok, I added support with the JobConfigs for `Require` and `Fulfill` so you can get an `EntitySpawner`.

I don't think the ACV approach really makes sense for an EntitySpawner since it is the underlying `Prototypes` hash map that is actually ACV'd. 

You can still use the System like usual with `SpawnDeferred` or `SpawnImmediate` for main thread spawning. Or you can `AcquireAsync` for custom job spawning.

TaskDrivers can easily use it via `RequireEntitySpawner`

Reply:
hmm, ok. The way the EntitySpawnWriter's adapted before was the same idea where the writer wasn't actually what required the access control it was the data that backed it.

All these special interfaces just create headaches whenever we need to use them in generic contexts (like JobConfig). I guess we can wait until I hit that some day and convert it then.

Reply:
Discussed and while not great, it's ok enough for now.